pop-api/src/error.rs

```diff
@@ -0,0 +1,405 @@
+	}
+}
+impl FromStatusCode for StatusCode {
+	fn from_status_code(status_code: u32) -> Result<(), Self> {
+		match status_code {
+			0 => Ok(()),
+			_ => {
+				let mut encoded = status_code.to_le_bytes();
+				convert_unknown_errors(&mut encoded);
+				Err(StatusCode::from(u32::from_le_bytes(encoded)))
+			},
+		}
+	}
+}
+
+impl From<scale::Error> for StatusCode {
+	fn from(_: scale::Error) -> Self {
+		DecodingFailed.into()
+	}
+}
+
+// If an unknown variant of the `DispatchError` is detected the error needs to be converted
+// into the encoded value of `Error::Other`. This conversion is performed by shifting the bytes one
+// position forward (discarding the last byte as it is not used) and setting the first byte to the
+// encoded value of `Other` (0u8). This ensures the error is correctly categorized as an `Other`
+// variant which provides all the necessary information to debug which error occurred in the runtime.
+//
+// Byte layout explanation:
+// - Byte 0: index of the variant within `Error`
+// - Byte 1:
+//   - Must be zero for `UNIT_ERRORS`.
+//   - Represents the nested error in `SINGLE_NESTED_ERRORS`.
+//   - Represents the first level of nesting in `DOUBLE_NESTED_ERRORS`.
+// - Byte 2:
+//   - Represents the second level of nesting in `DOUBLE_NESTED_ERRORS`.
+// - Byte 3:
+//   - Unused or represents further nested information.
+//
+// This mechanism ensures backward compatibility by correctly categorizing any unknown errors
+// into the `Other` variant, thus preventing issues caused by breaking changes.
+fn convert_unknown_errors(encoded_error: &mut [u8; 4]) {
+	let all_errors = [
+		UNIT_ERRORS.as_slice(),
+		SINGLE_NESTED_ERRORS.as_slice(),
+		DOUBLE_NESTED_ERRORS.as_slice(),
+		// `DecodingFailed`.
+		&[255u8],
+	]
+	.concat();
+	// Unknown errors, i.e. an encoded value where the first byte is non-zero (indicating a variant
+	// in `Error`) but unknown.
+	if !all_errors.contains(&encoded_error[0]) {
+		encoded_error[..].rotate_right(1);
+		encoded_error[0] = 0u8;
+	}
+	convert_unknown_nested_errors(encoded_error);
+}
+
+// If an unknown nested variant of the `DispatchError` is detected (i.e. when any of the subsequent
+// bytes are non-zero).
+fn convert_unknown_nested_errors(encoded_error: &mut [u8; 4]) {
+	// Converts single nested errors that are known to the Pop API as unit errors into `Other`.
+	if UNIT_ERRORS.contains(&encoded_error[0]) && encoded_error[1..].iter().any(|x| *x != 0u8) {
+		encoded_error[..].rotate_right(1);
+		encoded_error[0] = 0u8;
+	// Converts double nested errors that are known to the Pop API as single nested errors into
+	// `Other`.
+	} else if SINGLE_NESTED_ERRORS.contains(&encoded_error[0])
+		&& encoded_error[2..].iter().any(|x| *x != 0u8)
+	{
+		encoded_error[..].rotate_right(1);
+		encoded_error[0] = 0u8;
+	} else if DOUBLE_NESTED_ERRORS.contains(&encoded_error[0])
+		&& encoded_error[3..].iter().any(|x| *x != 0u8)
+	{
+		encoded_error[..].rotate_right(1);
+		encoded_error[0] = 0u8;
+	}
+}
```

Comment:
I am not a fan of this paradigm to use arrays to parse errors. It feels too much like C, when Rust's Enums are likely sufficient. For example, we can do a TryFrom<u8> to convert bytes into an Enum representation.

Although, one thing that may improve the current implementation and readability is to use pattern matching:
```
match encoded_error {
    [_, 0, 0, 0] => { /*non-nested error + validation here*/ },
    [_, _, 0, 0] => { /*single-nested error + validation here*/ },
    [_, _, _, 0] => { /*double-nested error + validation here*/ },
}
```

Reply:
https://github.com/r0gue-io/pop-node/compare/daan/feat-assets...daan/show_peter?expand=1

Reply:
Personally, I do like that more as it's easier to understand what's happening.

Reply:
Created a separate issue #91